Small changes to OMPS nadir mapper level 2 reader

mbruckner-work opened the pull request:
In testing the fix to the Melodies Monet driver I came across a small change I made to the OMPS reader that allows for using a list of files instead of expanding a string with glob to get the satellite files to read in. This change is necessary for processing over time intervals in Melodies Monet.

Commit:
small fixes to OMPS O3 level 2 reader (_omps_nadir_mm.py)

Commit:
corrected error handling in OMPS O3 level 2 reader

mbruckner-work committed:
remove debugging prints from omps_nadir

zmoon commented on monetio/sat/_omps_nadir_mm.py:
```diff
@@ -13,30 +13,36 @@ def read_OMPS_nm(files):
-
+            try:
+                data = extract_OMPS_nm(filename)
+                if count == 0:
+                    data_array = data
+                    count += 1
+                else:
+                    data_array = xr.concat([data_array, data], "x")
```
Can you move this block to an `else:` branch below the `except ...:`? This way `count` is only incremented if the try succeeds.

mbruckner-work replied:
`count` functions as a flag for initializing the combined dataset for all observations to be read in. Unless I'm mistaken, in its current location it only increments for the first file successfully read in.

zmoon replied:
Ah yes I see what you're saying. I would still move this stuff since it's really the `data = extract_OMPS_nm(filename)` that you're `try`ing and want to handle exceptions for. A better option though might be to have a list of data arrays that you append to upon successful load, then just call xarray concat once at the end. Then `count` isn't needed.

zmoon replied:
But fine to leave it this way too.

mbruckner-work replied:
@zmoon: I think that this may need to stay how it is until after the tutorial. The ValueError that comes up for files with weird cross-track dimensions occurs in the xarray concat, so appending on successful load and then calling xarray concat at the very end results in passing no data out of the reader if there is a ValueError.

zmoon replied:
Would you mind adding some comments about the `KeyError` and `ValueError` situations under the `except` line? Just short versions of your comment here about the `ValueError` and the one [here](https://github.com/noaa-oar-arl/monetio/pull/199#discussion_r1788230853).

mbruckner-work replied:
Added.

https://github.com/mbruckner-work/monetio/blob/8c027520e7ec61f111aa524594abc129a0cd4401/monetio/sat/_omps_nadir_mm.py#L39-L42

zmoon replied:
Great, thanks

zmoon commented on monetio/sat/_omps_nadir_mm.py:
```diff
@@ -13,30 +13,36 @@ def read_OMPS_nm(files):
+            except ValueError as e:
+                print(f'ValueError occured for: {filename}')
+                print(e)
     return data_array
```
```suggestion
    if count == 0:
        raise RuntimeError(f"no files loaded from files={files}")
    return data_array
```
Without something like this, `data_array` would be unbound if all the tries gave `KeyError` or `ValueError` and were skipped (or if `filelist` was empty), which would give a more confusing error message.

mbruckner-work replied:
Thanks for the suggestion. This is likely a more informative place to let the user know there were no files loaded.

zmoon commented on monetio/sat/_omps_nadir_mm.py:
```diff
@@ -13,30 +13,36 @@ def read_OMPS_nm(files):
 
     count = 0
-    print(files)
+    #print(files)
```
```suggestion
```

zmoon reviewed (commented):
Curious why you are expecting some files to fail to load. Does `extract_OMPS_nm` need modifications?

mbruckner-work committed:
remove an unnecessary print

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

zmoon commented on monetio/sat/_omps_nadir_mm.py:
```diff
@@ -13,30 +13,35 @@ def read_OMPS_nm(files):
+                    count += 1
+                else:
+                    data_array = xr.concat([data_array, data], "x")
+            except KeyError as e:
+                print(f'KeyError occured for: {filename}')
+                print(e)
+            except ValueError as e:
+                print(f'ValueError occured for: {filename}')
+                print(e)
```
```suggestion
            except (KeyError, ValueError) as e:
                print(f"warning: skipping {filename}. {type(e).__name__} occured: {e}")
```
I thought I had this in my review but it seems to have disappeared

mbruckner-work replied:
Yeah, I went to test a clean install and got error messages for reading in my downloaded test file set that I would have already dealt with.

The KeyError exception exists because I've run into files that contain no data and are essentially just a filename. The output message is "Unable to synchronously open object (object 'GeolocationData' doesn't exist)".   
I also came across an OMPS observation file where the y-dimension (satellite cross track) was very wrong for some reason and should be skipped.

mbruckner-work committed:
condense OMPS-NM level 2 error statement

Condense exceptions and associated error statements, improved information in error statement

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

mbruckner-work committed:
code comments added to _omps_nadir_mm.py

mbruckner-work committed:
Add error message when no files were loaded

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

mbruckner-work committed:
Add KeyError/ValueError situation comments

zmoon commented on monetio/sat/_omps_nadir_mm.py:
```diff
@@ -13,30 +13,35 @@ def read_OMPS_nm(files):
+                    data_array = xr.concat([data_array, data], "x")
+            except (KeyError, ValueError) as e:
+                # KeyError occurs when file exists but contains no data
+                # ValueError occurs when file cross-track dimensions are different than other files loaded
```
Correct?
```suggestion
                # KeyError occurs in load when file exists but contains no data
                # ValueError occurs in concat when file cross-track dimensions are different than other files loaded
```

mbruckner-work replied:
Yes.

mbruckner-work committed:
Update error comments

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

zmoon commented:
@mbruckner-work I can go ahead and resolve the linter/formatter stuff if you want.

mbruckner-work commented:
@zmoon That would be great. Thanks!

zmoon committed:
sp

oops sorry that was me..

zmoon committed:
fmt

zmoon reviewed (approved):
@mbruckner-work If you're satisfied, I will merge this. It's just the ISH tests failing afaict.
